### PR TITLE
[WIP]Fix cli export

### PR DESCRIPTION
**Pull request by takezyou**
##### SUMMARY

Bug fix export and import for AWX CLI.

related #7447 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

 - CLI

##### AWX VERSION

```
13.1.0
```

### CODE REVIEW

File: awxkit/awxkit/api/pages/inventory.py [hunk 1]
```diff
@@ -32,7 +32,7 @@
 class Inventory(HasCopy, HasCreate, HasInstanceGroups, HasVariables, base.Base):
 
     dependencies = [Organization]
-    NATURAL_KEY = ('organization', 'name')
+    NATURAL_KEY = ('organization', 'hosts', 'groups', 'name')
```
Comment (jbradberry) [inline awxkit/awxkit/api/pages/inventory.py]: Hosts and groups should not be part of the natural key here.  The natural key is supposed to be a (mostly) fixed minimal subset of the information of the object needed to uniquely look it up.

File: awxkit/awxkit/api/pages/inventory.py [hunk 2]
```diff
@@ -353,7 +354,7 @@ def remove_group(self, group):
 
 class Groups(page.PageList, Group):
 
-    pass
+    NATURAL_KEY = ('group',)
```
Comment (jbradberry) [inline awxkit/awxkit/api/pages/inventory.py]: I'm pretty sure 'group' isn't a field or related on the Group model.  You should check the Django models under the awx/ subtree of this repo to get a sense of what fields are used for the uniqueness constraint for this model.

File: awxkit/awxkit/api/pages/inventory.py [hunk 3]
```diff
@@ -434,7 +436,7 @@ def create(
 
 class Hosts(page.PageList, Host):
 
-    pass
+    NATURAL_KEY = ('host',)
```
Comment (jbradberry) [inline awxkit/awxkit/api/pages/inventory.py]: As with Groups, I don't believe 'host' is a field on Host.

File: awxkit/awxkit/api/pages/inventory.py (hunk 2 above)
Comment (jbradberry) [reply awxkit/awxkit/api/pages/inventory.py]: Also, I just noticed that this is on Groups.  The plural page type shouldn't get the natural key, only the singular version.

**Review by jbradberry — CHANGES_REQUESTED** [review]

**Comment by jbradberry** [comment]
I'm afraid that this latest change has moved further away from where things probably should be.  Here's what I think:

- [ ] `Host` and `Group` should be in `EXPORTABLE_DEPENDENT_OBJECTS`
- [ ] they should _not_ be in `EXPORTABLE_RESOURCES`
- [ ] they should be in `EXPORTABLE_RELATIONS`
- [ ] looking at awx/main/models these are both `unique_together` on 'name' plus 'inventory', these should be the fields used for the `natural_key`

This would export the objects only as part of the related objects on exported inventories, and would create them in-place on that inventory during import, plus hooking up the relationships between them.

With all of that said, though, I'm not sure what would need to be done about inventory sources, inventory scripts, and special inventories.  Maybe @chrismeyersfsu can comment about this?

**Comment by takezyou** [comment]
This ed6d68c commit can import but `maximum recursion depth exceeded while calling a Python object`  error.

**Comment by ryanpetrello** [comment]
Hey @takezyou,

Thanks for taking a stab at this.  When you're done tinkering, let's please squash all these commits into a single one - it looks like there was a lot of activity here.

**Comment by jladdjr** [comment]
hi @takezyou - I tried kicking the tires on these changes and found that while I can now see hosts in my export, the import isn't restoring the hosts. So, looks like we may be missing the second half.

For reference, here's the new export data that I'm seeing:
image


> With all of that said, though, I'm not sure what would need to be done about inventory sources, inventory scripts, and special inventories. 

^ as @jbradberry mentioned, inventories do touch on several other resources, though. Maybe, as a first step, we could have a PR that covers inventory basics (inventories, groups, hosts), and then grow that to include scripts, credentials, venvs, smart inventories, etc. There's a lot of ground to cover here.

**Comment by jladdjr** [comment]
@takezyou - TLDR: looks like issue w/ exporting / importing hosts is tied to the underlying export code itself, and not the changes made in this PR.

---

@jbradberry and I took a look at this and found that the export logic doesn't seem to be handling the case of hosts correctly. Here, specifically: https://github.com/ansible/awx/blob/devel/awxkit/awxkit/api/pages/api.py#L109-L114

The issue is that exported hosts -- like this one:

image

.. are missing a `"natural_key"` section. That section is a cue to the export code that the JSON block represents _an object that should be created_, instead of just being a reference to the object.

I think @jbradberry can explain this much better than I can, but from what I gathered, things seem tricky because of variations on how endpoints work; some allow creating resources and also associating resources, some are restricted to just creating the resource. Apparently, the inventory / host case falls into one category, while WFJTs and WFJTNs fall into the other. Sounds like the code block I mentioned had one case in mind, but not the other.

**Comment by takezyou** [comment]
@jladdjr Thank you comment!
I was able to notice the same problem when I was debugging it.

> I think @jbradberry can explain this much better than I can, but from what I gathered, things seem tricky because of variations on how endpoints work; some allow creating resources and also associating resources, some are restricted to just creating the resource. Apparently, the inventory / host case falls into one category, while WFJTs and WFJTNs fall into the other. Sounds like the code block I mentioned had one case in mind, but not the other.

@jbradberry What do you think about this?

**Comment by jbradberry** [comment]
@takezyou I have nothing further to add than what @jladdjr said.

**Comment by takezyou** [comment]
@jbradberry Please review.

**Comment by jbradberry** [comment]
@takezyou I think that the underlying issue needs to be addressed, rather than bolting on support on a case-by-case basis.

**Comment by jbradberry** [comment]
As noted, solving this is trickier than the approach outlined above.  Closing this in favor of #8368.

**Pull request closed without merging**